ray-operator/test/e2e/rayjob_suspend_test.go

```diff
@@ -0,0 +1,95 @@
+package e2e
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	. "github.com/ray-project/kuberay/ray-operator/test/support"
+)
+
+func TestRayJobSuspend(t *testing.T) {
+	test := With(t)
+
+	// Create a namespace
+	namespace := test.NewTestNamespace()
+
+	// Job scripts
+	jobs := newConfigMap(namespace.Name, "jobs", files(test, "long_running.py"))
+	jobs, err := test.Client().Core().CoreV1().ConfigMaps(namespace.Name).Create(test.Ctx(), jobs, metav1.CreateOptions{})
+	test.Expect(err).NotTo(HaveOccurred())
+	test.T().Logf("Created ConfigMap %s/%s successfully", jobs.Namespace, jobs.Name)
+
+	test.T().Run("Suspend the RayJob when its status is 'Running', and then resume it.", func(t *testing.T) {
+		// RayJob
+		rayJob := &rayv1.RayJob{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: rayv1.GroupVersion.String(),
+				Kind:       "RayJob",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "long-running",
+				Namespace: namespace.Name,
+			},
+			Spec: rayv1.RayJobSpec{
+				RayClusterSpec:           newRayClusterSpec(mountConfigMap(jobs, "/home/ray/jobs")),
+				Entrypoint:               "python /home/ray/jobs/long_running.py",
+				ShutdownAfterJobFinishes: true,
+				TTLSecondsAfterFinished:  600,
+				SubmitterPodTemplate:     jobSubmitterPodTemplate(),
+			},
+		}
+		rayJob, err = test.Client().Ray().RayV1().RayJobs(namespace.Name).Create(test.Ctx(), rayJob, metav1.CreateOptions{})
+		test.Expect(err).NotTo(HaveOccurred())
+		test.T().Logf("Created RayJob %s/%s successfully", rayJob.Namespace, rayJob.Name)
+
+		test.T().Logf("Waiting for RayJob %s/%s to be 'Running'", rayJob.Namespace, rayJob.Name)
+		test.Eventually(RayJob(test, rayJob.Namespace, rayJob.Name), TestTimeoutMedium).
+			Should(WithTransform(RayJobDeploymentStatus, Equal(rayv1.JobDeploymentStatusRunning)))
+
+		// Refresh the RayJob status
+		rayJob = GetRayJob(test, rayJob.Namespace, rayJob.Name)
+
+		test.T().Logf("Suspend the RayJob %s/%s", rayJob.Namespace, rayJob.Name)
+		rayJob.Spec.Suspend = true
+		// TODO (kevin85421): We may need to retry `Update` if 409 conflict makes the test flaky.
+		rayJob, err = test.Client().Ray().RayV1().RayJobs(namespace.Name).Update(test.Ctx(), rayJob, metav1.UpdateOptions{})
```

Comment:
A patch operation in that case could avoid dealing with conflicts.

Reply:
I am not familiar with `patch`. As I understand it, since it is not protected by `resourceVersion`, it could potentially lead to race conditions. Is it safe to use the `patch` operation in our scenario?

Reply:
That depends on the type of the patch request. Server-side apply has some conflict resolution mechanism relying on managed fields, but that's true for the others like JSON patch, merge patch, or strategic merge patch, there is no optimistic locking mechanisms in place (like checking `resourceVersion`). 

That being said, in our scenario here, I think it's safe as 1) the patch is localised to `.spec.suspend`, 2) the operator is not supposed to mutate the RayJob spec often, and even if/when is does, it would either do an update operation that would fail because of `resourceVersion` mismatch, or do patch operations on fields other than `.spec.suspend`.